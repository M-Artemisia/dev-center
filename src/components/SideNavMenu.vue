<template>
  <div ref="menu">
    <slot :toggle="toggle" :openSection="openSection"></slot>
  </div>
</template>

<script>
export default {
  name: 'SideNavMenu',
  props: {
    menuIdentifier: {
      type: String,
      required: true
    }
  },
  data() {
    return {
      openSection: ''
    }
  },
  methods: {
    toggle(sectionName) {
      this.openSection = this.openSection === sectionName ? '' : sectionName
    },
    getOpenSection() {
      const openLink = this.$refs.menu.querySelector(`li.cloak-open-nav-item[nav-menu-identifier="${this.menuIdentifier}"]`)
      if (openLink) {
        this.openSection = openLink.id.replace('-list-item', '')
      }
    },
  },
  mounted() {
    this.getOpenSection()
  }
}
</script>
