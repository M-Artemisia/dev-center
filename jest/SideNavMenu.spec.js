import { mount } from '@vue/test-utils'
import SideNavMenu from '@/components/SideNavMenu'

describe('SideNavMenu', () => {
  test('should find open list item and store in data', () => {
    const wrapper = mount(SideNavMenu, {
      propsData: {
        menuIdentifier: 'main'
      },
      slots: {
        default: `<ul>
  <li class="cloak-closed-nav-item" id="devcenter-list-item" nav-menu-identifier="main"></li>
  <li class="cloak-open-nav-item" id="api-list-item" nav-menu-identifier="main"></li>
  <li class="cloak-closed-nav-item" id="home-list-item" nav-menu-identifier="main"></li>
</ul>`
      }
    })
    expect(wrapper.vm.$data.openSection).toEqual('api')
  })

  test('should update active section when toggle method called by slot', () => {
    const wrapper = mount(SideNavMenu, {
      propsData: {
        menuIdentifier: 'main'
      },
      scopedSlots: {
        default: `<ul slot-scope="{ toggle }">
  <li @click="toggle('devcenter')" class="cloak-closed-nav-item" id="devcenter-list-item" nav-menu-identifier="main"></li>
  <li @click="toggle('api')" class="cloak-open-nav-item" id="api-list-item" nav-menu-identifier="main"></li>
  <li class="cloak-closed-nav-item" id="home-list-item" nav-menu-identifier="main"></li>
</ul>`
      }
    })
    wrapper.find('#devcenter-list-item').trigger('click')
    expect(wrapper.vm.$data.openSection).toEqual('devcenter')
  })

  test('should pass open section as a slot prop', () => {
    const wrapper = mount(SideNavMenu, {
      propsData: {
        menuIdentifier: 'main'
      },
      scopedSlots: {
        default: `<ul slot-scope="{ openSection }">
  <li class="cloak-closed-nav-item" id="devcenter-list-item" nav-menu-identifier="main" :class="{'open-nav-item': openSection === 'devcenter'}"></li>
  <li class="cloak-open-nav-item" id="api-list-item" nav-menu-identifier="main" :class="{'open-nav-item': openSection === 'api'}"></li>
  <li class="cloak-closed-nav-item" nav-menu-identifier="main" id="home-list-item"></li>
</ul>`
      }
    })
    expect(wrapper.html()).toMatchSnapshot()
  })
})
