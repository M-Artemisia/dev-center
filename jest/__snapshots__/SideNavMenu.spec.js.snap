// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SideNavMenu should pass open section as a slot prop 1`] = `
<div>
  <ul>
    <li id="devcenter-list-item" nav-menu-identifier="main" class="cloak-closed-nav-item"></li>
    <li id="api-list-item" nav-menu-identifier="main" class="cloak-open-nav-item open-nav-item"></li>
    <li nav-menu-identifier="main" id="home-list-item" class="cloak-closed-nav-item"></li>
  </ul>
</div>
`;
